apps/devhub/tests/test_utils.py

```diff
@@ -0,0 +1,373 @@
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': True,
+                          'matched': True})
+
+            self.compare({'id': ('a', 'b', 'c'),
+                          'signing_severity': severity,
+                          'ignore_duplicates': False,
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': False,
+                          'matched': True})
+
+        # Other severities are ignored only when flagged as ignorable.
+        for severity in 'medium', 'high':
+            self.compare({'id': ('a', 'b', 'c'),
+                          'signing_severity': severity,
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': False,
+                          'matched': True})
+
+            self.compare({'id': ('a', 'b', 'c'),
+                          'signing_severity': severity,
+                          'ignore_duplicates': True,
+                          'context': ('x', 'y', 'z'),
+                          'file': 'foo.js'},
+                         {},
+                         {'ignored': True,
+                          'matched': True})
+
+        # Check that unchecked properties don't matter.
+        self.compare({'id': ('a', 'b', 'c'),
+                      'signing_severity': 'low',
+                      'context': ('x', 'y', 'z'),
+                      'file': 'foo.js'},
+                     {'thing': 'stuff'},
+                     {'ignored': True,
+                      'matched': True})
+
+        # Non-signing messages should be matched, but not annotated for
+        # ignorability.
+        self.compare({'id': ('a', 'b', 'c'),
+                      'context': ('x', 'y', 'z'),
+                      'file': 'foo.js'},
+                     {},
+                     {'matched': True})
+
+    def test_non_matching_messages(self):
+        """Test that messages which should match, don't."""
+
+        message = {'id': ('a', 'b', 'c'),
+                   'signing_severity': 'low',
+                   'context': ('x', 'y', 'z'),
+                   'context_data': {'function': 'foo_bar'},
+                   'file': 'foo.js'}
+
+        self.compare(message, {'id': ('d', 'e', 'f')}, {})
+        self.compare(message, {'signing_severity': 'high'}, {})
+        self.compare(message, {'context': ('a', 'b', 'c')}, {})
+        self.compare(message, {'context_data': {}}, {})
+        self.compare(message, {'file': 'these-are-not-the-droids.js'}, {})
+
+        # Token non-signing message.
+        self.compare({'id': ('a', 'b', 'c'),
+                      'context': ('x', 'y', 'z'),
+                      'file': 'foo.js'},
+                     {'id': ()}, {})
```

Comment:
I think I'd rather add tests for the test code than try to expand it. I initially wrote a few tests with complete versions of original, modified, and expected messages, and it was so noisy that it wasn't obvious what was being changed between messages, so I wasn't confident that they were correct/would stay correct. And repeatedly creating, updating, and testing messages in the same way seems less likely to be reliable than writing the code once and using it for all of the tests that test this behavior.